mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,468 @@
+        startOffset - lastToken.map(_.startOffset).getOrElse(0)
+      else
+        startOffset - line.startOffset
+    }
+    def countCR: Unit = { crCount += 1 }
+  }
+
+  /**
+   * This function returns -1 when capable Type is nothing.
+   *  TokenTypes that can be on multilines are handled in another func.
+   *  See Token.Comment in this file.
+   */
+  private def typeOfNonIdentToken(
+      tk: scala.meta.tokens.Token
+  ): Integer = {
+    tk match {
+      // case _: Token.Ident => // in case of Ident is
+
+      // Alphanumeric keywords
+      case _: Token.ModifierKeyword => getTypeId(SemanticTokenTypes.Modifier)
+      case _: Token.Keyword => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwNull => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwTrue => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwFalse => getTypeId(SemanticTokenTypes.Keyword)
+
+      // extends Symbolic keywords
+      case _: Token.Hash => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Viewbound => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.LeftArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Subtype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.RightArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Supertype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.At => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Underscore => getTypeId(SemanticTokenTypes.Variable)
+      case _: Token.TypeLambdaArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.ContextArrow => getTypeId(SemanticTokenTypes.Operator)
+
+      // Constant
+      case _: Token.Constant.Int | _: Token.Constant.Long |
+          _: Token.Constant.Float | _: Token.Constant.Double =>
+        getTypeId(SemanticTokenTypes.Number)
+      case _: Token.Constant.String | _: Token.Constant.Char =>
+        getTypeId(SemanticTokenTypes.String)
+      case _: Token.Constant.Symbol => getTypeId(SemanticTokenTypes.Property)
+
+      // Comment
+      case _: Token.Comment => getTypeId(SemanticTokenTypes.Comment)
+
+      // Interpolation
+      case _: Token.Interpolation.Id | _: Token.Interpolation.SpliceStart =>
+        getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Interpolation.Start | _: Token.Interpolation.Part |
+          _: Token.Interpolation.SpliceEnd | _: Token.Interpolation.End =>
+        getTypeId(SemanticTokenTypes.String) // $ symbol
+
+      // Default
+      case _ => -1
+    }
+
+  }
+  def pickFromTraversed(tk: scala.meta.tokens.Token): Option[NodeInfo] = {
+
+    val adjustForBacktick: Int = {
+      var ret: Int = 0
+      val cName = tk.text.toCharArray()
+      if (cName.size >= 2) {
+        if (
+          cName(0) == 96.toChar // backtick
+          && cName(cName.size - 1) == 96.toChar
+        ) ret = 2
+      }
+      ret
+    }
+
+    val buffer = ListBuffer.empty[NodeInfo]
+    for (node <- nodes) {
```

Comment:
Here we always go over all the nodes and if the file is big we get `O(n^2)` complexity. Instead since we are going linearly over positions we could have nodes as a sorted list and then only use a mutable iterator.

Something along the lines of:
```
val nodeIterator = nodes.iterator
var currentNode =  nodesIterator.next()
... 
while (currentNode.pos.start < tk.pos.start && nodesIterator.hasNext()) node = nodesIterator.next()
if (
        currentNode.pos.start == tk.pos.start &&
        currentNode.pos.end + adjustForBacktick == tk.pos.end
      ) Some(currentNode)
else
  None
```
This way we go over the nodes collection once.

One complication is that you would need to keep `nodeIterator` and `currentNode` between the invocations of `pickFromTraversed`